maturin 1.3 produces invalid source distributions for workspace packages

### Bug Description

Hi folks!

First off, thank y'all for building maturin. It's been super pleasant and smooth sailing so far. 🍰 
We've hit a small roadbump with 1.3.0 and 1.3.1 which I wanted to report. :)

```
$ maturin build --release --compatibility manylinux2014 --zig --sdist
$ pip install target/wheels/*.tar.gz
Processing ./target/wheels/mitmproxy_rs-0.4.0.tar.gz
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'error'
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [13 lines of output]
      error: failed to load manifest for workspace member `/tmp/pip-req-build-_e2p_y8n/mitmproxy-windows/redirector`
      
      Caused by:
        failed to read `/tmp/pip-req-build-_e2p_y8n/mitmproxy-windows/redirector/Cargo.toml`
      
      Caused by:
        No such file or directory (os error 2)
      💥 maturin failed
        Caused by: Cargo metadata failed. Does your crate compile with `cargo build`?
        Caused by: `cargo metadata` exited with an error:
      Error running maturin: Command '['maturin', 'pep517', 'write-dist-info', '--metadata-directory', '/tmp/pip-modern-metadata-aoszm7an', '--interpreter', '/opt/hostedtoolcache/Python/3.11.6/x64/bin/python']' returned non-zero exit status 1.
      Checking for Rust toolchain....
      Running `maturin pep517 write-dist-info --metadata-directory /tmp/pip-modern-metadata-aoszm7an --interpreter /opt/hostedtoolcache/Python/3.11.6/x64/bin/python`
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

Notice:  A new release of pip is available: 23.2.1 -> 23.3.1
Notice:  To update, run: pip install --upgrade pip
Error: Process completed with exit code 1.
```

### Your maturin version (`maturin --version`)

1.3.1

### Your Python version (`python -V`)

Python 3.12

### Your pip version (`pip -V`)

pip 23.2.1

### What bindings you're using

pyo3

### Does `cargo build` work?

- [X] Yes, it works

### If on windows, have you checked that you aren't accidentally using unix path (those with the forward slash `/`)?

- [X] Yes

### Steps to Reproduce

1. Update maturin from 1.2.3 to 1.3.1
2. Build an sdist for a package in a workspace.

maturin 1.2.3:
```
$ ls -l mitmproxy_rs-0.3.11/
-rwxrwxrwx 1 user user  68K Sep 28 05:11 Cargo.lock
-rwxrwxrwx 1 user user  985 Jan  1  1970 Cargo.toml
-rwxrwxrwx 1 user user 3.7K Jan  1  1970 PKG-INFO
-rwxrwxrwx 1 user user 2.8K Sep 28 05:11 README.md
drwxrwxrwx 1 user user 4.0K Nov  6 11:37 local_dependencies
drwxrwxrwx 1 user user 4.0K Nov  6 11:37 mitmproxy_rs
-rwxrwxrwx 1 user user  827 Sep 28 05:11 pyproject.toml
drwxrwxrwx 1 user user 4.0K Nov  6 11:37 src
-rwxrwxrwx 1 user user  166 Sep 28 05:11 stubtest-allowlist.txt
```

maturin 1.3.1:
```
$ ls -l mitmproxy_rs-0.4.0/
-rwxrwxrwx 1 user user 6.1K Nov  4 11:12 CHANGELOG.md
-rwxrwxrwx 1 user user 2.0K Nov  4 11:12 CONTRIBUTING.md
-rwxrwxrwx 1 user user  68K Nov  4 11:12 Cargo.lock
-rwxrwxrwx 1 user user 2.2K Nov  4 11:12 Cargo.toml
-rwxrwxrwx 1 user user 1.1K Nov  4 11:12 LICENSE
-rwxrwxrwx 1 user user 3.7K Jan  1  1970 PKG-INFO
-rwxrwxrwx 1 user user 3.1K Nov  4 11:12 README.md
-rwxrwxrwx 1 user user 238K Nov  4 11:12 architecture.png
drwxrwxrwx 1 user user 4.0K Nov  6 11:37 benches
drwxrwxrwx 1 user user 4.0K Nov  6 11:37 mitmproxy-macos
drwxrwxrwx 1 user user 4.0K Nov  6 11:37 mitmproxy-rs
drwxrwxrwx 1 user user 4.0K Nov  6 11:37 mitmproxy-windows
drwxrwxrwx 1 user user 4.0K Nov  6 11:37 mitmproxy_rs
-rwxrwxrwx 1 user user  883 Jan  1  1970 pyproject.toml
drwxrwxrwx 1 user user 4.0K Nov  6 11:37 src
```

```
$ ls -l mitmproxy_rs-0.4.0/mitmproxy-windows/
-rwxrwxrwx 1 user user  555 Nov  4 11:12 README.md
drwxrwxrwx 1 user user 4.0K Nov  6 11:37 mitmproxy_windows
-rwxrwxrwx 1 user user  596 Nov  4 11:12 pyproject.toml
```

It looks like `mitmproxy-windows/redirector` is missing entirely from the sdist, even though it's a workspace member.

See https://github.com/mitmproxy/mitmproxy_rs/pull/123 for a full reproducer. :)

--- a/src/source_distribution.rs
+++ b/src/source_distribution.rs
@@ -148,6 +148,7 @@ fn add_crate_to_source_distribution(
     prefix: impl AsRef<Path>,
     known_path_deps: &HashMap<String, PathDependency>,
     root_crate: bool,
+    skip_cargo_toml: bool,
 ) -> Result<()> {
     let manifest_path = manifest_path.as_ref();
     let output = Command::new("cargo")
@@ -216,7 +217,7 @@ fn add_crate_to_source_distribution(
     if root_crate {
         let rewritten_cargo_toml = rewrite_cargo_toml(manifest_path, known_path_deps)?;
         writer.add_bytes(cargo_toml_path, rewritten_cargo_toml.as_bytes())?;
-    } else {
+    } else if !skip_cargo_toml {
         writer.add_file(cargo_toml_path, manifest_path)?;
     }
 
@@ -345,6 +346,10 @@ fn add_cargo_package_files_to_sdist(
     let workspace_manifest_path = workspace_root.join("Cargo.toml");
 
     let known_path_deps = find_path_deps(&build_context.cargo_metadata)?;
+    debug!(
+        "Found path dependencies: {:?}",
+        known_path_deps.keys().collect::<Vec<_>>()
+    );
     let mut sdist_root =
         common_path_prefix(workspace_root.as_std_path(), pyproject_toml_path).unwrap();
     for path_dep in known_path_deps.values() {
@@ -361,21 +366,31 @@ fn add_cargo_package_files_to_sdist(
 
     // Add local path dependencies
     for (name, path_dep) in known_path_deps.iter() {
+        debug!(
+            "Adding path dependency: {} at {}",
+            name,
+            path_dep.manifest_path.display()
+        );
         let path_dep_manifest_dir = path_dep.manifest_path.parent().unwrap();
         let relative_path_dep_manifest_dir =
             path_dep_manifest_dir.strip_prefix(&sdist_root).unwrap();
+        // we may need to rewrite workspace Cargo.toml later so don't add it to sdist yet
+        let skip_cargo_toml = workspace_manifest_path == path_dep.manifest_path;
         add_crate_to_source_distribution(
             writer,
             &path_dep.manifest_path,
             &root_dir.join(relative_path_dep_manifest_dir),
             &known_path_deps,
             false,
+            skip_cargo_toml,
         )
-        .context(format!(
-            "Failed to add local dependency {} at {} to the source distribution",
-            name,
-            path_dep.manifest_path.display()
-        ))?;
+        .with_context(|| {
+            format!(
+                "Failed to add local dependency {} at {} to the source distribution",
+                name,
+                path_dep.manifest_path.display()
+            )
+        })?;
         // Handle possible relative readme field in Cargo.toml
         if let Some(readme) = path_dep.readme.as_ref() {
             let abs_readme = path_dep_manifest_dir
@@ -417,6 +432,7 @@ fn add_cargo_package_files_to_sdist(
         root_dir.join(relative_main_crate_manifest_dir),
         &known_path_deps,
         true,
+        false,
     )?;
     // Handle possible relative readme field in Cargo.toml
     if let Some(readme) = main_crate.readme.as_ref() {

